[quant][graphmode][fx] Add support for dynamic quant for RNN and RNNCell

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49179 [quant][be] Add typing for quantization_mappings.py
* **#49126 [quant][graphmode][fx] Add support for dynamic quant for RNN and RNNCell**

Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_rnn
python test/test_quantization.py TestQuantizeFxOps.test_rnn_cell

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25449047](https://our.internmc.facebook.com/intern/diff/D25449047)

jerryzh168 committed:
[quant][graphmode][fx] Add support for dynamic quant for RNN and RNNCell

Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_rnn
python test/test_quantization.py TestQuantizeFxOps.test_rnn_cell

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for dynamic quant for RNN and RNNCell"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_rnn
python test/test_quantization.py TestQuantizeFxOps.test_rnn_cell

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -2107,6 +2111,48 @@ def forward(self, indices, offsets):
+            }
+            model_graph = prepare_fx(model_graph, graph_qconfig_dict)
+            model_graph = convert_fx(model_graph)
+            self.assertEqual(model_eager(sample_input), model_graph(sample_input))
```
Do we need to test for serialization here?

jerryzh168 replied:
this should be the same as eager mode module, I'm not too familiar, are we using state_dict?

jerryzh168 replied:
or are you referring to checkScriptable

jerryzh168 replied:
added checkScriptable here, but in general we'll do e2e test in TestQuantizeFxModels

vkuzo commented on torch/quantization/quantization_mappings.py:
```diff
@@ -124,6 +124,19 @@ def get_static_quant_module_class(float_module_class, additional_static_quant_ma
         " does not have a corresponding quantized module class"
     return static_quant_module_class
 
+def get_dynamic_quant_module_class(float_module_class, additional_dynamic_quant_mapping=None):
```
would be great to add types to function I/O

jerryzh168 replied:
we can add it in a separate PR I think, all other functions in this file are not typed yet

vkuzo replied:
it's not blocking this PR, but would be awesome if we started adding these as we go, at least to function I/O.  We don't have to wait for a file to have existing type annots to add more.  This also distributes the cost of adding them to everyone, as opposed to one person.

jerryzh168 replied:
yeah, I'm saying I plan to add it in a separate PR, or are you suggesting to add the type annotations for the functions in this file in this PR?

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -491,6 +492,48 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
+            return quantizer.quantized_graph.node_copy(node, load_arg(quantized=None))
+
+        module = quantizer.modules[node.target]
+        qmodule = get_dynamic_quant_module_class(type(module))
+        quantized = qmodule.from_float(module)
```
optional nits: how about `qmodule_cls`, since it's not a class instance, and `quantized_mod`, since `quantized` is used in L535-536?

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for dynamic quant for RNN and RNNCell"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_rnn
python test/test_quantization.py TestQuantizeFxOps.test_rnn_cell

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25449047](https://our.internmc.facebook.com/intern/diff/D25449047)

[ghstack-poisoned]
